packages/stream-api/src/webSocketClient.ts

```diff
@@ -95,11 +81,41 @@ export class WebSocketClient {
     }
   }
 
+  private onError(reject: VoidFunction): void {
+    this.removeWebSocketListeners();
+    reject();
+  }
+
+  private onClose(): void {
+    this.removeWebSocketListeners();
+    if (this.urlProvider != null) {
+      this.handleClose(this.urlProvider);
+    }
+  }
+
+  private removeWebSocketListeners(): void {
+    if (this.webSocket != null) {
+      this.webSocket.removeEventListener('message', this.handleMessage);
+      this.webSocket.removeEventListener('open', () =>
+        this.onOpen(Promise.resolve)
+      );
+      this.webSocket.removeEventListener('error', () =>
+        this.onError(Promise.reject)
+      );
```

Comment:
This won't actually remove the listeners, and will result in a memory leak. You need to pass the same reference to the function that passed to `addEventListener`.

Reply:
Here's what I'd recommend doing. Define a function that closes over the function references and return a disposable that can later be called to clean up the WS.

```ts
  private addWebSocketListeners = (
    ws: WebSocket,
    resolve: VoidFunction,
    reject: VoidFunction
  ): Disposable => {
    const onOpen = (): void => this.onOpen(resolve);
    const onError = (): void => reject();
    ws.addEventListener('message', this.handleMessage);
    ws.addEventListener('open', onOpen);
    ws.addEventListener('error', onError);
    ws.addEventListener('close', this.onClose);

    return {
      dispose: () => {
        ws.removeEventListener('message', this.handleMessage);
        ws.removeEventListener('open', onOpen);
        ws.removeEventListener('error', onError);
        ws.removeEventListener('close', this.onClose);
      },
    };
  };

  private removeWebSocketListeners(): void {
    this.listeners?.dispose();
  }
```

Reply:
Thanks!